src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs

```diff
@@ -69,9 +69,22 @@ private DataType FromJson(JObject json)
             return this;
         }
 
-        internal override bool NeedConversion() => true;
+        internal override bool NeedConversion() => ElementType.NeedConversion();
 
-        internal override object FromInternal(object obj) => throw new NotImplementedException();
+        internal override object FromInternal(object obj)
```

Comment:
Do we have a test that captures `Collect` with array types?

Reply:
This [ udf test](https://github.com/dotnet/spark/blob/e162ff8731f9cb53668a807ca421daa24f2b9723/src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs#L52) should cover `Collect`.